src/mlpack/tests/distribution_test.cpp

```diff
@@ -442,6 +442,97 @@ BOOST_AUTO_TEST_CASE(GammaDistributionTrainTest)
   BOOST_REQUIRE_EQUAL(gDist.Dimensionality(), d2);
 }
 
+BOOST_AUTO_TEST_CASE(GammaDistributionTrainWithProbabilitiesTest)
+{
+  double alphaReal = 5.4;
+  double betaReal = 6.7;
+
+  // Create a gamma distribution random generator
+  std::default_random_engine generator;
+  std::gamma_distribution<double> dist(alphaReal, betaReal);
+
+  size_t N = 500;
+  size_t d = 2;
+  arma::mat rdata(d, N);
+
+  for(size_t j = 0; j < d; j++)
+    for(size_t i = 0; i < N; i++)
+      rdata(j, i) = dist(generator);
+
+  // create a uniform distribution random generator
+  std::uniform_real_distribution<double> prob(0, 1);
+  arma::vec probabilities(N);
+
+  for(size_t i = 0; i < N; i++)
+    probabilities(i) = prob(generator);
+
+  // fit results with probabilities and data
+  GammaDistribution gDist;
+  gDist.Train(rdata, probabilities);
+
+  // fit results with only data
+  GammaDistribution gDist2;
+  gDist2.Train(rdata);
+
+  BOOST_REQUIRE_CLOSE(gDist2.Alpha(0), gDist.Alpha(0), 10);
+  BOOST_REQUIRE_CLOSE(gDist2.Beta(0), gDist.Beta(0), 10);
+
+  BOOST_REQUIRE_CLOSE(alphaReal, gDist.Alpha(0), 10);
+  BOOST_REQUIRE_CLOSE(betaReal, gDist.Beta(0), 10);
+
+
+  // This test ensures that the same result is obtained when
+  // trained with probabilities all set to 1 and with no probabilities at all.
+
+  // fit results with data and each probability as 1
+  GammaDistribution gDist3;
+  arma::vec all_probabilities_1(N, arma::fill::ones);
+  gDist3.Train(rdata, all_probabilities_1);
+
+  BOOST_REQUIRE_EQUAL(gDist2.Alpha(0), gDist3.Alpha(0));
+  BOOST_REQUIRE_EQUAL(gDist2.Beta(0), gDist3.Beta(0));
+
+  // This test draws points from two different gamma distributions,
+  // sets the probabilities for the points from the first distribution
+  // to something small and the probabilities for the second to something large. 
+  // It ensures that the gamma distribution recovered has the
+  // same parameters as the second gamma distribution with high probabilities.
+
+  double alphaReal2 = 1.9;
+  double betaReal2 = 8.4;
+
+  std::uniform_real_distribution<double> low_prob(0, 0.02);
+  std::uniform_real_distribution<double> high_prob(0.98, 1);
+
+  std::gamma_distribution<double> dist2(alphaReal2, betaReal2);
+
+  // draws points alternately from the two different distributions.
+  for(size_t j = 0; j < d; j++)
+  {
+    for(size_t i = 0; i < N; i++)
+    {
+      if(i%2 == 0)
+        rdata(j, i) = dist(generator);
+      else
+        rdata(j, i) = dist2(generator);
+    }
+  }
+
+  for(size_t i = 0; i<N; i++)
+  {
+    if(i%2 == 0)
+      probabilities(i) = low_prob(generator);
+    else
+      probabilities(i) = high_prob(generator);
+  }
```

Comment:
This would require ensuring shuffling rdata matrix and probabilities vector in the same order. How to use shuffle for this?